nomad/alloc_endpoint.go

```diff
@@ -445,3 +445,138 @@ func (a *Alloc) GetServiceRegistrations(
 		},
 	})
 }
+
+// GetIdentities allows nodes to retrieve workload identities for their
+// allocations.
+//
+// This is an internal-only RPC and not exposed via the HTTP API.
+func (a *Alloc) GetIdentities(args *structs.AllocIdentitiesRequest, reply *structs.AllocIdentitiesResponse) error {
+
+	authErr := a.srv.Authenticate(a.ctx, args)
+
+	// Ensure the connection was initiated by a client if TLS is used.
+	if err := validateTLSCertificateLevel(a.srv, a.ctx, tlsCertificateLevelClient); err != nil {
+		return err
+	}
+	if done, err := a.srv.forward("Alloc.GetIdentities", args, args, reply); done {
+		return err
+	}
+	a.srv.MeasureRPCRate("alloc", structs.RateMetricRead, args)
+	if authErr != nil {
+		return structs.ErrPermissionDenied
+	}
+
+	defer metrics.MeasureSince([]string{"nomad", "alloc", "get_identities"}, time.Now())
+
+	if len(args.Identities) == 0 {
+		// Client bug
+		return fmt.Errorf("no identities requested")
+	}
+
+	opts := blockingOptions{
+		queryOpts: &args.QueryOptions,
+		queryMeta: &reply.QueryMeta,
+		run: func(ws memdb.WatchSet, state *state.StateStore) error {
+			// Lookup the allocations
+			thresholdMet := false
+			for _, idReq := range args.Identities {
+				out, err := state.AllocByID(ws, idReq.AllocID)
+				if err != nil {
+					return err
+				}
+
+				if out == nil {
+					// Alloc may have been GC'd and therefore should not be able to get
+					// new identities.
+					continue
+				}
+
+				if out.ModifyIndex > args.QueryOptions.MinQueryIndex {
+					thresholdMet = true
+					break
+				}
+				a.logger.Debug("threshold not met", "min", args.QueryOptions.MinQueryIndex, "max", out.ModifyIndex)
+			}
+
+			// If we could not find an alloc updated after the desired index, note
+			// the index allocs were last updated.
+			if !thresholdMet {
+				index, err := state.Index("allocs")
+				if err != nil {
+					return err
+				}
+				reply.Index = index
+				return nil
+			}
+
+			// Threshold met, so create the response
+			now := time.Now().UTC()
+			maxIndex := uint64(0)
+			for _, idReq := range args.Identities {
+				out, err := state.AllocByID(ws, idReq.AllocID)
```

Comment:
Each blocking `run` function call operates on a snapshot (if I'm understanding the code in `rpc.go` correctly), so the allocations we got above should be unchanged by the time we get here. Can we reuse them? Should we add to the list of rejections when we first try to get the allocation?

Reply:
Yes! You made me realize I was looking up the alloc for each identity, despite the current implementation only sending identities for a single alloc in a single request... so we'd just load the same alloc over and over. :facepalm: 

I fixed that and then reused the allocs we look up.

The only potential downside I see is that this code now diverges wildly from other blocking queries. I didn't do an exhaustive search but most if not all existing blocking queries seem to have `return blockingRPC(...)` at the end and stuff _all_ of their post-auth/post-validation logic into that blockingRPC callback.

I don't see a reason this is a problem, but after so many times copying and pasting (:grimacing:) the same RPC code around it does feel funny to go my own way. The way the callback mutates some captured variables for use after the blocking callback (`allocs` and `thresholdMet`) feels icky, but I think that's my only complaint.

In fact this split should make it far easier to reuse the bottom identity creation bits in `Alloc.GetAllocs` when we implement the "create alternate identities when allocs are first fetched" optimization!

(Sorry for the long ramble. Even though the tests over this are fairly comprehensive I'm really having to talk myself into straying from the well trod `return blockingRPC(...)` path.)